codex/stores/maintenance.nim

```diff
@@ -10,90 +10,318 @@
+
+        if currDatasetMd.leavesCount <= datasetMd.checkpoint.progress:
+          DatasetMetadata.none
+        else:
+          datasetMd.some
+      else:
+        raise newException(CatchableError, "Metadata for dataset with treeCid " & $treeCid & " not found")
   )
 
-  without iter =? expirations, err:
-    trace "Unable to obtain blockExpirations iterator from repoStore"
-    return
+proc deleteBatch(self: DatasetMaintainer, treeCid: Cid, datasetMd: DatasetMetadata): Future[?!void] {.async.} =
+  var datasetMd = datasetMd
+
+  datasetMd.checkpoint.timestamp = await self.clock.now
+
+  if err =? (await self.recordCheckpoint(treeCid, datasetMd)).errorOption:
+    return failure(err)
+
+  logScope:
+    treeCid = treeCid
+    manifestsCids = datasetMd.manifestsCids
+    startingIndex = datasetMd.checkpoint.progress
+
+  trace "Deleting a batch of blocks", size = self.batchSize
+
+  var index = datasetMd.checkpoint.progress
+  while (index < datasetMd.checkpoint.progress + self.batchSize) and
+          (index < datasetMd.leavesCount):
+    if err =? (await self.blockStore.delBlock(treeCid, index)).errorOption:
+      error "Error deleting a block", msg = err.msg, index
+
+    index.inc
+
+    await sleepAsync(50.millis) # cooperative scheduling
+
+  if index >= datasetMd.leavesCount:
+    trace "All blocks deleted from a dataset", leavesCount = datasetMd.leavesCount
+
+    for manifestCid in datasetMd.manifestsCids:
+      if err =? (await self.blockStore.delBlock(manifestCid)).errorOption:
+        error "Error deleting manifest", cid = manifestCid
+
+    if err =? (await self.blockStore.delBlock(treeCid)).errorOption:
+      error "Error deleting block", cid = treeCid
+
+    if err =? (await self.recordCheckpoint(treeCid, datasetMd)).errorOption:
+      return failure(err)
+
+    return success()
+  else:
+    datasetMd.checkpoint.progress = index
+    return await self.deleteBatch(treeCid, datasetMd)
+
+proc superviseDatasetDeletion(self: DatasetMaintainer, treeCid: Cid, datasetMd: DatasetMetadata): Future[void] {.async.} =
+  logScope:
+    treeCid = treeCid
+    manifestsCids = datasetMd.manifestsCids
+    expiry = datasetMd.expiry
+    leavesCount = datasetMd.leavesCount
+
+  try:
+    if datasetMd.checkpoint.progress == 0 and datasetMd.checkpoint.timestamp == 0:
+      info "Initiating deletion of a dataset"
+    else:
+      info "Retrying deletion of a dataset", progress = datasetMd.checkpoint.progress, timestamp = datasetMd.checkpoint.timestamp
+
+    if err =? (await self.deleteBatch(treeCid, datasetMd)).errorOption:
+      error "Error occurred during deletion of a dataset", msg = err.msg
+    else:
+      info "Dataset deletion complete"
+  except CatchableError as err:
+    error "Unexpected error during dataset deletion", msg = err.msg, treeCid = treeCid
+
+proc listDatasetMetadata*(
+  self: DatasetMaintainer
+): Future[?!AsyncIter[?!(Cid, DatasetMetadata)]] {.async.} =
```

Comment:
That simple change is more complicated than it seems - even though I agree with it and I would like to see it. 

This change requires doing filtering of items which are futures. `AsyncIter` has a property `finished: bool` that indicates if there are other items left to read. The implicit contract is that it's always safe to fetch next item if `finished = false`. However it's impossible to tell beforehand if the iterator has finished or not if we do filtering of values that will be resolved in future. 

This is a flaw in `AsyncIterator` interface, but `AsyncIter` was just a quick and dirty way to remove frequently occurring duplicated code that already had this flaw.

Reply:
I'm not sure I understand. The `AsyncIter` is returning a `Result[(Cid, DatasetMetadata)]`. My suggestion was to change it so that `AsyncIter` returns `CidDatasetMetadata` to make this more readable. This should not impact how the AsyncIter functions with filters.

My point was that returning this Result type from the AsyncIter is not needed, because there's only one case that's a failure. Looking at the callsite of `listDatasetMetadata`, we see that in the case of an error, it is simply logged. Therefore, if the error was logged immediately (on line 292), there is no need to return a Result from the AsyncIter, thus allowing us to simplify the return type and make it more readable.

Reply:
Ok, I think I will have a solution to that, and I will implement it. Thanks for persisting on fixing this.

To put you in my perspective, imagine that you have an iterator that returns `Future[?!T]` and you would like to change this iterator to return `Future[T]`. You need to filter out all `failure(err)` items and unwrap `success(value)` items into just `value` and then yield them. Let's say all `Future[?!T]` will resolve to be `failure(err)` but not immediately, they'll resolve after 10 seconds. So if you filter out all `failute(err)` values you end up with an empty (finished) iterator, but you know that only after 10 seconds has passed. What should be reported by the `finished` flag during 0-9 seconds period?

I will solve it, by doing mapping `Iter[Future[?!T]] -> Future[Iter[Future[T]]]`, so that the whole iterator is not resolved immediately but after some await time: either await for next item (finished=false) or for draining it and finding out there are no items (finished=true). I hope it'll work.

Reply:
This particular routine, `listDatasetMetadata` is complicated because of the way `queryT` is built. My opinion is that we do not need `queryT` at all. 

Please look more closely at `codex/sales/reservations`, as it it is doing almost exactly what you're describing, and it uses generics to get strongly typed results. Here we can see the usage of `storables` and `items` lumped into the `all` routine, such that a consumer could call `reservations.all(Availability)` and get all Availabilities in the datastore (asynchronously) by returning a `Future[?!seq[T]]`.
https://github.com/codex-storage/nim-codex/blob/916e4834d088c9e1837b0a30e8403095fd8b565e/codex/sales/reservations.nim#L494-L516

where `items` is a simple iterator:
https://github.com/codex-storage/nim-codex/blob/916e4834d088c9e1837b0a30e8403095fd8b565e/codex/sales/reservations.nim#L453-L455

and `storables` builds up the async iterator with a filter:
https://github.com/codex-storage/nim-codex/blob/916e4834d088c9e1837b0a30e8403095fd8b565e/codex/sales/reservations.nim#L457-L492

You can ignore the namespaces.len filter above, as this is a weak design and currently being changed.

I'm more than happy to discuss this on a call, so we can maybe figure out a good solution together. Let me know if you're up for that 👍

Reply:
> What should be reported by the finished flag during 0-9 seconds period?

You might be able to use an `AsyncEvent` for `finished` instead of a `bool`, and possibly a second flag like `querying` as a simple bool. The iterator would continue to loop while `querying` is true, and it would wait for `finished` to be fired (blocking the coroutine asynchronously). Once all items have been queried, `querying` would be set to false and the iterator would finish.

Reply:
Ok, I reviewed it and it looks quite similar, but it's currently scoped in `reservations.nim` and uses `json` as a codec instead of a generic coder `T->seq[byte]`. I think it would be probably better to refactor `reservations.nim` to use shared methods from a shared module like this `typedatastore.nim` than to repeat the same approach over multiple modules.

Reply:
> uses json as a codec instead of a generic coder T->seq[byte]

The point I was trying to make is that the *pattern* used in `reservations` can be used in the `maintenance` module. Once results are fetched from the datastore, the decoder (whether that be a json deserializer or protobuf decoder) can decode the bytes from the datastore and return strongly typed results. 

> I think it would be probably better to refactor reservations.nim to use shared methods from a shared module like this typedatastore.nim than to repeat the same approach over multiple modules.

I disagree because sometimes refactoring for DRY purposes makes things much more complex, difficult to read, understand, and therefore maintain. The benefit to using a pattern versus an ultra-generic "fits all use cases" module is that a pattern is easy to read, and it can be modified for a particular use case (eg there may be something needed in the `maintenance` pattern that is not needed in the `reservations` pattern, like an `AsyncEvent` for `finished` in the iterator, as an example). Whereas a generic, all-purpose DRY module must allow for these unique cases, which creates too much complexity. In the `reservations` pattern, it provides strongly typed results without the need for an overly complex and hard-to-read `typeddatastore` module.

Reply:
>I disagree because sometimes refactoring for DRY purposes makes things much more complex, difficult to read, understand, and therefore maintain.

I agree with principle that DRY doesn't always lead to better code, but this is not one of these cases. The single responsibility of `typedatastore.nim` module is to handle encoding and decoding the items from `seq[byte]` to `T` and vice versa. That's all. Just adding this module cuts a lot of boilerplate from `repostore.nim` and `maintanance.nim`, and would cut it from `reservations.nim` as well.

All of the cases in `repostore.nim` and `maintanence.nim` could be covered and are covered by abstractions in `typeddatastore.nim`. If we will encounter a case that will not be covered by them, there's nothing stopping us from using the approach that you propose, which is using the Datastore interface directly and doing encoding/decoding manually.